### ISSUE TITLE
If still exec the compress process when inference?

### ISSUE THREAD

**Issue by as754770178** [Tencent/PocketFlow#214]
I read the code in `export_pb_tflite_models.py`. The following code compress the model, but I think these operation will save in the `pb` file, and the new output `x` rely on `kernel_gthr`, `kernel_shrk ` and `conv` operation. I want to know these operation will exec when inference?
```
      kernel_gthr = np.zeros((1, 1, kernel_chn_in, nnzs.size))
      kernel_gthr[0, 0, nnzs, np.arange(nnzs.size)] = 1.0
      kernel_shrk = kernel[:, :, nnzs, :] 
      x = tf.nn.conv2d(op.inputs[0], kernel_gthr, [1, 1, 1, 1], 'SAME', data_format=data_format)
      x = tf.nn.conv2d(
        x, kernel_shrk, strides, padding, data_format=data_format, dilations=dilations)
```

**Comment by as754770178** [comment]
thanks.  
1. Whether the model size will be reduced? Because the model add new variable, such as, kernel_gthr and kernel_shrk, and not delete the old variable.
2. Whether the model calculation will be reduced? Because add new operation, such as
```
      x = tf.nn.conv2d(op.inputs[0], kernel_gthr, [1, 1, 1, 1], 'SAME', data_format=data_format)
      x = tf.nn.conv2d(
        x, kernel_shrk, strides, padding, data_format=data_format, dilations=dilations)
```

**Comment by jiaxiang-wu** [comment]
1. Old variables are deleted in *.pb & *.tflite models, since they are not used when computing the final outputs.
2. If the kernel size is larger than 1 (*e.g.* 3x3 conv), the FLOPs will be reduced. For 1x1 conv, we recommend using the "tf.gather + 3 x 3 conv" scheme for graph transformation.

**Comment by as754770178** [comment]
This is the pb file generated by `export_pb_tflite_models.py`, that the pb file of compress model is bigger than the pb file of original model.
```
-rw-rw-r--. 1 zgz zgz  95082914 Jan 25 15:53 model_original.pb
-rw-rw-r--. 1 zgz zgz 104845311 Jan 25 15:53 model_transformed.pb
```

**Comment by jiaxiang-wu** [comment]
Which model are you using?

**Comment by as754770178** [comment]
I use the model save in `cp_best_path`.

**Comment by jiaxiang-wu** [comment]
For ResNet-50 model, the 1x1 conv in each bottleneck structure cannot be compressed with the "1x1 conv + 3x3 conv" scheme for graph transformation. For 1x1 conv, we recommend using the "tf.gather + 3 x 3 conv" scheme.

**Comment by as754770178** [comment]
1. The conv in residual brach will be compressed? and why the size of pb file become bigger?
2. How to use the "tf.gather + 3 x 3 conv" scheme?

**Comment by jiaxiang-wu** [comment]
1. The conv ops in the residual branch are compressed, but the size of *.pb file can sometimes be larger. Assume a 1x1 conv layer with 64 input channels is pruned to 32, then # of original parameters is 1*1*64*c (c: # of output channels). When using the "1x1 conv + 3x3 conv" scheme, this layer will be decomposed into two layers: one 1x1 conv with 64 input channels and 32 output channels and one 1x1 conv with 32 input channels and c output channels. So, the overall # of parameters is 1*1*64*32+1*1*32*c. This can be larger than the original one if c is no larger than 64.
2. Are you using the latest version of PocketFlow? Or, can you find the following code in `export_pb_tflite_models.py`?
``` Python
      # replace channel pruned convolutional with cheaper operations
      if graph_trans_mthd == 'gather':
        x = tf.gather(op.inputs[0], nnzs, axis=1)
        x = tf.nn.conv2d(
          x, kernel_shrk, strides, padding, data_format=data_format, dilations=dilations)
      elif graph_trans_mthd == '1x1_conv':
        x = tf.nn.conv2d(op.inputs[0], kernel_gthr, [1, 1, 1, 1], 'SAME', data_format=data_format)
        x = tf.nn.conv2d(
          x, kernel_shrk, strides, padding, data_format=data_format, dilations=dilations)
      else:
        raise ValueError('unrecognized graph transformation method: ' + graph_trans_mthd)
```

**Comment by as754770178** [comment]
sorry, my code is not the latest version

**Comment by jiaxiang-wu** [comment]
Also, take a look at this PR #119 
![](https://user-images.githubusercontent.com/12831717/49414783-89eabd00-f7af-11e8-8c14-8553d95fc15a.png)

**Comment by as754770178** [comment]
The "tf.gather + 3 x 3 conv" scheme is only for `NCHW`?
`graph_trans_mthd = 'gather' if data_format == 'NCHW' else '1x1_conv'`

**Comment by as754770178** [comment]
1. The inference speed  of the pruned model is `sample/sec: 72+`, The inference speed of the pruned model is `sample/sec: 53+`.  This does not match the description in the PocketFlow documentation. The `cp_uniform_preserve_ratio` is 0.5, the inference speed of `pruned model` is 4 times that of `original model`.

2. Can use `tf.gather(op.inputs[0], nnzs, axis=3)` in `NHWC` format model?

**Comment by jiaxiang-wu** [comment]
@as754770178 
`tf.gather` can be used in `NHWC` format models, but this is less efficient than that in `NCHW` format models.
